### PR TITLE
Coordinate Transformations

### PR DESCRIPTION
**Pull request by gschulze**
This PR implements all coordinate transformation methods used in GeoTIFF:

- Tie Point and Pixel Scale
- Affine Transform
- Multiple Tie Points (Ground Control Points)

In addition, the implementation considers the raster type (`RasterPixelIsArea` or `RasterPixelIsPoint`) to determine the origin of the raster space coordinate system.

The _Tie Point and Pixel Scale_ and _Affine Transform_ methods are relatively straightforward. The _Multiple Tie Points_ transformation roughly works as follows:

- Create a Delaunay triangulation of the tie points in raster space
- Apply the resulting triangulation to the tie points in model space
- When doing coordinate transformations, use the containing triangle of a coordinate to determine the three points to be used for the transformation
- Determine the relative position of the coordinate within the triangle in the model space coordinate system
- Use the relative position to linearly interpolate the coordinate within the corresponding triangle in the raster space coordinate system

@weiji14 please have a look.

### CODE REVIEW

File: Cargo.toml
Comment (kylebarron) [inline Cargo.toml:17]: Since in this PR it looks like you're only using an rtree statically (e.g. you don't need to insert items after the initial bulk load), you may want to consider a static index like https://github.com/kylebarron/geo-index, which I've benchmarked to be [reasonably faster](https://github.com/kylebarron/geo-index#benchmarks) and much more memory efficient than rtree.

File: src/coordinate_transform/affine_transform.rs
Comment (kylebarron) [inline src/coordinate_transform/affine_transform.rs:14]: Slight nit: since it doesn't look like you actually mutate `transform` anywhere after initialization, I find it a bit easier to read if you use
```rs
let transform = [
transformation_matrix[0],
transformation_matrix[1],
transformation_matrix[3],
transformation_matrix[4],
transformation_matrix[5],
transformation_matrix[7]
];
```
because then I don't have to guess whether `transform` will be mutated or not later on.

(Ditto for `inverse_transform`)

Comment (weiji14) [reply src/coordinate_transform/affine_transform.rs:14]: Side tangent: I'm thinking if we could actually just use https://docs.rs/geo/0.28.0/geo/algorithm/affine_ops/struct.AffineTransform.html instead of re-implementing things here.

Comment (gschulze) [reply src/coordinate_transform/affine_transform.rs]: > Slight nit: since it doesn't look like you actually mutate `transform` anywhere after initialization, I find it a bit easier to read if you use
> 
> ```rust
> let transform = [
> transformation_matrix[0],
> transformation_matrix[1],
> transformation_matrix[3],
> transformation_matrix[4],
> transformation_matrix[5],
> transformation_matrix[7]
> ];
> ```
> 
> because then I don't have to guess whether `transform` will be mutated or not later on.
> 
> (Ditto for `inverse_transform`)

Agree, changed in 23b7ac0ec38e284e5e6112b1222e97b4bb47a70b.

Comment (gschulze) [reply src/coordinate_transform/affine_transform.rs]: > Side tangent: I'm thinking if we could actually just use https://docs.rs/geo/0.28.0/geo/algorithm/affine_ops/struct.AffineTransform.html instead of re-implementing things here.

I thought about it too but decided against it to avoid including ´geo´ as a dependency. I'm not sure whether the costs outweigh the benefits.

Comment (weiji14) [reply src/coordinate_transform/affine_transform.rs]: > > Side tangent: I'm thinking if we could actually just use https://docs.rs/geo/0.28.0/geo/algorithm/affine_ops/struct.AffineTransform.html instead of re-implementing things here.
> 
> I thought about it too but decided against it to avoid including ´geo´ as a dependency. I'm not sure whether the costs outweigh the benefits.

Yeah, I feel like the affine transform ops could be split out into its own thing (the gdal crate could probably use some of this too). Ok to keep things as is for now, and maybe keep an eye on https://github.com/georust/geo/issues/1186 if `geo-types` ever merges with `geo`.

Comment (gschulze) [reply src/coordinate_transform/affine_transform.rs]: Ok, will mark this one as resolved, then.

File: Cargo.toml
Comment (weiji14) [inline Cargo.toml:14]: Thoughts on putting these behind a feature flag, considering that they're only used for GeoTIFF with GCPs (that are relatively rare)?

File: tests/integration.rs
Comment (weiji14) [inline tests/integration.rs:260]: Ok if we move the `test_transform_*` unit tests to a separate file? (I've also been thinking of renaming `tests/integration.rs` to something else, but can do that afterwards).

**Review by weiji14 — COMMENTED** [review]
Just a couple of quick comments on the organization of the code. I'll try to find some time to do a deeper dive over the weekend!

File: Cargo.toml
Comment (gschulze) [reply Cargo.toml:14]: Yes, makes sense. What should it be called, "gcp"?

File: tests/integration.rs
Comment (gschulze) [reply tests/integration.rs:260]: `tests/transform.rs`?

File: Cargo.toml
Comment (weiji14) [reply Cargo.toml:14]: I'm not very good with names :sweat_smile: Maybe "tiepoints"?

File: tests/integration.rs
Comment (weiji14) [reply tests/integration.rs:260]: Looking at https://doc.rust-lang.org/book/ch11-03-test-organization.html#submodules-in-integration-tests, is it the convention to use `integration_` as the prefix for integration tests in Rust?

Comment (gschulze) [reply tests/integration.rs:260]: No, I don't think so. The file is simply called ´integration_test.rs´ to highlight that it is an integration test as opposed to a unit test.

**Comment by kylebarron** [comment]
I'm not familiar with how GCPs work, so if @weiji14 has time to review I'll defer to him

File: src/geo_key_directory.rs
Comment (weiji14) [inline src/geo_key_directory.rs:671]: A couple of thoughts here:

1. Should we add enum variants for `RasterType=0` (undefined/unknown) and `RasterType=32767` (user-defined), even though they're not recommended at https://docs.ogc.org/is/19-008r4/19-008r4.html#_requirements_class_gtrastertypegeokey?
2. Could you add a link to the GeoTIFF spec's `GTRasterTypeGeoKey` requirements for this please

File: src/coordinate_transform.rs
Comment (weiji14) [inline src/coordinate_transform.rs:22]: Document this by adding a link to https://docs.ogc.org/is/19-008r4/19-008r4.html#_raster_to_model_coordinate_transformation_requirements?

**Review by weiji14 — COMMENTED** [review]
Did an initial look through the changes and looks good so far! Unit tests are solid (you're lucky I did a exchange in Salzvurg, Austria many years ago, so familiar with the geography 😃), and I do appreciate the changes to `get_pixel_value` to return Some/None.

Leaving a few small comments or now, I'd be keen on a bit more documentation around the public enums/structs if possible (a link to the relevant subsection in the GeoTIFF spec will do). Will squeeze out some time next week to do a deeper review on the affine and Delaunay triangulation math.

**Review by weiji14 — APPROVED** [review]
Ok, I've checked the affine transform code and it looks sound. The Delaunay triangulation logic is a bit over my head, but I trust that it is ok (if not, someone will hopefully notice and file a bug report).

Don't want to hold this back any longer because I'm getting into a busy period, so I'll pre-approve this. I have some questions below around the `get_value_at` method, but hopefully those are non-issues.

**Comment by gschulze** [comment]
Thanks @weiji14 for your comments. I've been busy the last two weeks, but hopefully, I'll find time to address the remaining issues next week.

File: src/coordinate_transform/tie_points.rs
Comment (chuckwondo) [inline src/coordinate_transform/tie_points.rs:272]: Perhaps you meant this instead?

```rust
                if c.x < min_x {
                    min_x = c.x;
                }
                if c.x > max_x {
                    max_x = c.x;
                }
                if c.y < min_y {
                    min_y = c.y;
                }
                if c.y > max_y {
                    max_y = c.y;
                }
```

or even this?

```rust
                min_x = min_x.min(c.x);
                max_x = max_x.max(c.x);
                min_y = min_y.min(c.y);
                max_y = max_y.max(c.y);
```

File: src/geo_key_directory.rs
Comment (gschulze) [reply src/geo_key_directory.rs:671]: Done.

File: src/coordinate_transform.rs
Comment (weiji14) [reply src/coordinate_transform.rs:22]: Hmm, I haven't seen any new commits yet, did you do a `git push`?

Comment (gschulze) [reply src/coordinate_transform.rs:25]: I got interrupted today, sorry. Pushed what I've done so far.

File: src/lib.rs
Comment (gschulze) [reply src/lib.rs:115]: Added comments for public methods.

File: src/lib.rs
Comment (gschulze) [reply src/lib.rs:155]: @weiji14 The 0.5 offset refers to half a pixel, which only makes sense in raster space. See https://docs.ogc.org/is/19-008r4/19-008r4.html#_raster_space for reference. This issue also has some nice illustrations: https://github.com/opengeospatial/ogcapi-coverages/issues/92

I added a reference to the corresponding section in the standard in the code.

File: src/coordinate_transform/tie_points.rs
Comment (chuckwondo) [reply src/coordinate_transform/tie_points.rs]: @gschulze, are you sure this is resolved? You're comparing both `c.x` and `c.y` to _only_ `min_x` in _all_ cases. Am I missing something?

**Comment by gschulze** [comment]
@weiji14 Let me know if you want any further changes; otherwise, I will merge this PR.

File: src/lib.rs
Comment (weiji14) [reply src/lib.rs]: Ah ok, I got mixed up on raster space (which should be pixels) and [model space](https://docs.ogc.org/is/19-008r4/19-008r4.html#_model_space) (which are real world coordinates in metres or other units). I kept thinking raster offset should be in metres and model offset is in pixels, but it should be the other way around. Thanks for providing the links for clarification, and for adding an inline comment to document this in the code!

**Comment by weiji14** [comment]
Let's go ahead and merge, thanks for adding the extra docs and bug fixes!

**Pull request merged**
